### PR TITLE
Inplace and non-padded encoding for ABIEncoderV2.

### LINKED ISSUE #4821: Support packed encoding in ABIEncoderV2
This is also required for indexed event parameters.

### PR DESCRIPTION
**Pull request by chriseth**
Fixes #4821 
Fixes https://github.com/ethereum/solidity/issues/5695

Replaces https://github.com/ethereum/solidity/pull/5812

### CODE REVIEW

**Comment by chriseth** [comment]
One thing that is missing is the packed encoding of an array of strings.

**Comment by chriseth** [comment]
Added test for array of strings.

**Comment by chriseth** [comment]
Rebased.

**Comment by Marenz** [comment]
Fixed CI failure caused by my broken test

File: libsolidity/codegen/CompilerUtils.cpp [hunk 1]
```diff
@@ -348,11 +348,15 @@ void CompilerUtils::encodeToMemory(
 
 	if (_givenTypes.empty())
 		return;
-	else if (_padToWordBoundaries && !_copyDynamicDataInPlace && encoderV2)
+	if (encoderV2)
 	{
 		// Use the new Yul-based encoding function
+		solAssert(
+			!_padToWordBoundaries == _copyDynamicDataInPlace,
```
Comment (ekpyron) [inline libsolidity/codegen/CompilerUtils.cpp]:
```suggestion
			_padToWordBoundaries != _copyDynamicDataInPlace,
```

File: libsolidity/codegen/ABIFunctions.cpp [hunk 2]
```diff
@@ -107,6 +107,73 @@ string ABIFunctions::tupleEncoder(
...
+	TypePointers const& _targetTypes
+)
+{
+	EncodingOptions options;
+	options.encodeAsLibraryTypes = false;
+	options.encodeFunctionFromStack = true;
+	options.padded = false;
+	options.dynamicInplace = true;
+
+	string functionName = string("abi_encode_tuple_");
```
Comment (ekpyron) [inline libsolidity/codegen/ABIFunctions.cpp]: Will this name potentially collide with the non-packed version? Even if not, maybe still better to be safe:
```suggestion
	string functionName = string("abi_encode_tuple_packed_");
```

File: libsolidity/codegen/ABIFunctions.cpp [hunk 3]
```diff
@@ -702,15 +778,30 @@ string ABIFunctions::abiEncodingFunctionSimpleArray(
...
 		templ("dataAreaFun", arrayDataAreaFunction(_from));
-		templ("elementEncodedSize", toCompactHexWithPrefix(_to.baseType()->calldataEncodedSize()));
+		if (!usesTail)
+		{
+			if (_options.dynamicInplace && dynamicBase)
+			{
+				templ("assignToPos", "pos :=");
+				templ("incrementPos", "");
+			}
+			else
```
Comment (ekpyron) [inline libsolidity/codegen/ABIFunctions.cpp]: Ok, so this is the case of ``!usesTail && !(_options.dynamicInplace && dynamicBase)``..
So... ``!(dynamicBase && !_options.dynamicInplace) && !(_options.dynamicInplace && dynamicBase)``...
So... ``(!dynamicBase || _options.dynamicInplace) && (!_options.dynamicInplace || !dynamicBase)``
So... ``!dynamicBase``, right? Maybe that's worth a comment..

Comment (ekpyron) [reply libsolidity/codegen/ABIFunctions.cpp]: Maybe it's easier to start with that case as ``if (!dynamicBase)`` and have an ``else if(_options.dynamicInPlace)`` which should be the other case above? Not sure, but the logic is a bit hard to read as it is.

**Review by ekpyron — COMMENTED** [review]
Some initial comments, but I made the mistake to look at the entire thing at once, which is quite hard - I need to take another look commit by commit.

File: libsolidity/codegen/ABIFunctions.cpp (hunk 2 above)
Comment (chriseth) [reply libsolidity/codegen/ABIFunctions.cpp]: `options.toFunctionNameSuffix` should take care of this, but perhaps it does not hurt to add it in addition to that.

**Comment by axic** [comment]
> Some initial comments, but I made the mistake to look at the entire thing at once, which is quite hard - I need to take another look commit by commit.

Same here - it is a monster.

**Comment by chriseth** [comment]
The cases that have been discussed should be drastically simplified now.

**Comment by chriseth** [comment]
We can move the last commit to its own PR - it was just added because nobody started reviewing... :)

File: libsolidity/codegen/ABIFunctions.h [hunk 4]
```diff
@@ -134,6 +136,14 @@ class ABIFunctions
 		Type const& _targetType,
 		EncodingOptions const& _options
 	);
+	/// @returns the name of a function that internally calls `abiEncodingFunction`
+	/// but always returns the updated encoding position, even if the type is
+	/// statically encoded.
+	std::string abiEncodingFunctionUpdatedPos(
```
Comment (ekpyron) [inline libsolidity/codegen/ABIFunctions.h]: Not sure what's the best name for this. The current ``abiEncodingFunctionUpdatedPos`` might be ok, but maybe ``abiEncodeAndReturnUpdatedPosFunction``? Although thinking about it we might consider changing all the names in this class and remove the ``abi`` prefix and ``Function`` suffix, since due to them being in the outer class anyways, to me they seem redundant - then this could just be ``encodeAndReturnUpdatedFunction``, which would be much clearer I think - maybe something for another PR.

File: libsolidity/codegen/ABIFunctions.h [hunk 5]
```diff
@@ -232,9 +246,11 @@ class ABIFunctions
...
 
 	/// @returns the name of a function used during encoding that stores the length
-	/// if the array is dynamically sized. It returns the new encoding position.
-	/// If the array is not dynamically sized, does nothing and just returns the position again.
-	std::string arrayStoreLengthForEncodingFunction(ArrayType const& _type);
+	/// if the array is dynamically sized (and the options do not request in-place encoding).
+	/// It returns the new encoding position.
+	/// If the array is not dynamically sized (or in-place encoding was requested),
+	/// does nothing and just returns the position again.
+	std::string arrayStoreLengthForEncodingFunction(ArrayType const& _type, EncodingOptions const& _options);
```
Comment (ekpyron) [inline libsolidity/codegen/ABIFunctions.h:263]: This could at some point be renamed to ``storeLengthIfDynamicArray`` or just ``storeArrayLength``, but again probably fine for now and to be considered for another PR.

File: libsolidity/codegen/ABIFunctions.h [hunk 6]
```diff
@@ -212,6 +222,10 @@ class ABIFunctions
 	/// Pads with zeros and might write more than exactly length.
 	std::string copyToMemoryFunction(bool _fromCalldata);
 
+	/// @returns the name of a function that takes a (cleaned) value of the given type and left-aligns
+	/// it, usually for use in non-padded encoding.
```
Comment (ekpyron) [inline libsolidity/codegen/ABIFunctions.h]: Maybe this should document that this will assert/throw for non-value types? Might be obvious enough though...

File: libsolidity/codegen/ABIFunctions.cpp [hunk 7]
```diff
@@ -574,20 +574,66 @@ string ABIFunctions::abiEncodingFunction(
 			// special case: convert storage reference type to value type - this is only
```
Comment (ekpyron) [inline libsolidity/codegen/ABIFunctions.cpp:642]: Two lines above: ``to.isValueType()`` is asserted to be true here anyways (line 554), isn't it? Not an issue of this PR, but could still be changed here.

File: libsolidity/codegen/ABIFunctions.cpp [hunk 8]
```diff
@@ -819,7 +864,7 @@ string ABIFunctions::abiEncodingFunctionCompactStorageArray(
 			size_t storageBytes = _from.baseType()->storageBytes();
 			size_t itemsPerSlot = 32 / storageBytes;
 			// This always writes full slot contents to memory, which might be
-			// more than desired, i.e. it writes beyond the end of memory.
+			// more than desired, i.e. it alwayswrites beyond the end of memory.
```
Comment (ekpyron) [inline libsolidity/codegen/ABIFunctions.cpp]: Typo ``always writes``... but seems to be gone in subsequent commits.

File: libsolidity/codegen/ABIFunctions.cpp [hunk 9]
```diff
@@ -1007,9 +1070,10 @@ string ABIFunctions::abiEncodingFunctionStringLiteral(
 
 		if (_to.isDynamicallySized())
 		{
+			solAssert(_to.category() == Type::Category::Array, "");
 			Whiskers templ(R"(
 				function <functionName>(pos) -> end {
-					mstore(pos, <length>)
+					pos := <storeLength>(pos, <length>)
 					<#word>
 						mstore(add(pos, <offset>), <wordValue>)
```
Comment (ekpyron) [inline libsolidity/codegen/ABIFunctions.cpp:1146]: Do you think it'd be better to assign ``pos := add(pos, <wordSize>)`` after this ``mstore``, where ``<wordSize>`` is always 32 except for the last element in the non-padded case where it's ``value.size()&31``? Then below it could just be ``end := pos`` and we have less and smaller literal numbers in the code... But not strictly speaking related to this PR and we could keep it as is for now.

Not sure whether the optimizer would add everything up to the same offsets in the end anyways - but I guess that direction is easier to do later than going back to incrementing in small steps - I'm not sure.
All of that optional for this PR in any case.

**Review by ekpyron — COMMENTED** [review]
I reviewed the first commit and have some comments, but they are all optional, so in general the first commit looks fine to me for a start.
EDIT: everything is addressed in the now second commit.

File: libsolidity/codegen/ABIFunctions.cpp (hunk 9 above)
Comment (ekpyron) [reply libsolidity/codegen/ABIFunctions.cpp:1146]: Do we have/Should we have an issue for looking into the optimizer generating large constants by "simplifying" that might be more expensive than repeated increments?

File: libsolidity/codegen/ABIFunctions.h (hunk 4 above)
Comment (chriseth) [reply libsolidity/codegen/ABIFunctions.h]: Not all of the functions are actually strictly ABI related - the conversion functions are generic, for example. Will change the name of the new function.

File: libsolidity/codegen/ABIFunctions.h (hunk 5 above)
Comment (chriseth) [reply libsolidity/codegen/ABIFunctions.h:263]: The "encoding" should suggest that it is for storing the length during abi encoding and not e.g. for regular memory or storage arrays.

File: libsolidity/codegen/ABIFunctions.cpp (hunk 9 above)
Comment (chriseth) [reply libsolidity/codegen/ABIFunctions.cpp:1146]: The optimizer will add everything up anyway. I think the question is rather what is easier to verify in this source file. I don't think there is such an issue, and I'm also not sure whether we should create issues about all these things...

Comment (ekpyron) [reply libsolidity/codegen/ABIFunctions.cpp:1146]: Yeah... it's a lot of issues already... alright!

File: libsolidity/codegen/ABIFunctions.cpp (hunk 7 above)
Comment (ekpyron) [reply libsolidity/codegen/ABIFunctions.cpp:642]: Resolved in another PR.

**Comment by chriseth** [comment]
Argh, I just wanted to say it fixes a comment in this PR...

File: libsolidity/codegen/CompilerUtils.cpp (hunk 1 above)
Comment (ekpyron) [reply libsolidity/codegen/CompilerUtils.cpp]: I'd still suggest this :-). (EDIT: ``_padToWordBoundaries != _copyDynamicDataInPlace`` that is)

File: libsolidity/codegen/ABIFunctions.cpp [hunk 10]
```diff
@@ -107,6 +107,73 @@ string ABIFunctions::tupleEncoder(
...
+				string(R"(
+					pos := <abiEncode>(<values> pos)
+				)") :
+				string(R"(
+					<abiEncode>(<values> pos)
+					pos := add(pos, <calldataEncodedSize>)
+				)")
+			);
+			elementTempl("values", valueNames);
+			elementTempl("calldataEncodedSize", to_string(_targetTypes[i]->calldataEncodedSize(false)));
```
Comment (ekpyron) [inline libsolidity/codegen/ABIFunctions.cpp]: Could prepend ``if (dynamic)`` - doesn't matter too much, though.

**Review by ekpyron — COMMENTED** [review]
The now third commit ``Add packed encoder V2.`` looks good to me as well now.

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 11]
```diff
@@ -13715,25 +13715,23 @@ BOOST_AUTO_TEST_CASE(abi_encodePackedV2_structs)
...
+			function testMemory() public {
 				S memory m = s;
 				emit E(m);
-				return abi.encodePacked(uint8(0x33), m, uint8(0x44));
 			}
 		}
 	)";
 	compileAndRun(sourceCode, 0, "C");
 	bytes structEnc = encodeArgs(int(0x12), u256(-7), int(2), int(3), u256(-7), u256(-8));
 	string encoding = "\x33" + asString(structEnc) + "\x44";
```
Comment (ekpyron) [inline test/libsolidity/SolidityEndToEndTest.cpp]: Unused now, isn't it?

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 12]
```diff
@@ -13749,22 +13747,21 @@ BOOST_AUTO_TEST_CASE(abi_encodePackedV2_nestedArray)
...
 				x[1][0][1].a = 3;
 				x[1][1][1].b = 4;
 				emit E(x);
-				return abi.encodePacked(uint8(0x33), x, uint8(0x44));
 			}
 		}
 	)";
 	compileAndRun(sourceCode, 0, "C");
 	bytes structEnc = encodeArgs(1, 2, 3, 0, 0, 0, 0, 4);
 	string encoding = "\x33" + asString(structEnc) + "\x44";
```
Comment (ekpyron) [inline test/libsolidity/SolidityEndToEndTest.cpp]: Same

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 13]
```diff
@@ -13798,11 +13793,11 @@ BOOST_AUTO_TEST_CASE(abi_encodePackedV2_arrayOfStrings)
 	// This pads to multiple of 32 bytes
 	string encoding = "\x33" + asString(arrayEncoding) + "\x44";
```
Comment (ekpyron) [inline test/libsolidity/SolidityEndToEndTest.cpp]: Same - next line seems superfluous as well now.

**Review by ekpyron — COMMENTED** [review]
I think the rest is fine as well now apart from some minor comments. Only thing I haven't checked yet, is that the semantics tests are correct and exhaustive.

**Review by ekpyron — DISMISSED** [review]
I think this is fine now.

**Comment by chriseth** [comment]
Rebased and squashed..
